oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/MissingBcSweeper2.java

```diff
@@ -0,0 +1,280 @@
+            throws DocumentStoreException {
+        totalCount = 0;
+        lastCount = 0;
+        startOfScan = context.getClock().getTime();
+        lastLog = startOfScan;
+
+        Iterable<Map.Entry<Path, UpdateOp>> ops = sweepOperations(documents);
+        for (List<Map.Entry<Path, UpdateOp>> batch : partition(ops, INVALIDATE_BATCH_SIZE)) {
+            Map<Path, UpdateOp> updates = newHashMap();
+            for (Map.Entry<Path, UpdateOp> entry : batch) {
+                updates.put(entry.getKey(), entry.getValue());
+            }
+            listener.sweepUpdate(updates);
+        }
+        LOG.debug("Document sweep2 finished");
+    }
+
+    private Iterable<Map.Entry<Path, UpdateOp>> sweepOperations(
+            final Iterable<NodeDocument> docs) {
+        return filter(transform(docs,
+                new Function<NodeDocument, Map.Entry<Path, UpdateOp>>() {
+
+            int yieldCnt = 0;
+            long lastYield = System.currentTimeMillis();
+
+            @Override
+            public Map.Entry<Path, UpdateOp> apply(NodeDocument doc) {
+                if (++yieldCnt >= YIELD_SIZE) {
+                    try {
+                        Thread.sleep(Math.max(1, System.currentTimeMillis() - lastYield));
+                    } catch (InterruptedException e) {
+                        // ignore
+                    }
+                    lastYield = System.currentTimeMillis();
+                    yieldCnt = 0;
+                }
+                return immutableEntry(doc.getPath(), sweepOne(doc));
+            }
+        }), new Predicate<Map.Entry<Path, UpdateOp>>() {
+            @Override
+            public boolean apply(Map.Entry<Path, UpdateOp> input) {
+                return input.getValue() != null;
+            }
+        });
+    }
+
+    private UpdateOp sweepOne(NodeDocument doc) throws DocumentStoreException {
+        UpdateOp op = null;
+        // the blueprint NodeDocumentSweeper.sweepOne goes through
+        // PROPERTY_OR_DELETED_OR_COMMITROOT here, as that's the new full sweep1.
+        // the sweep2 though, only goes through COMMITROOT_OR_REVISIONS,
+        // as that's what was left out by the original sweep1:
+        // - COMMITROOT : for new child (parent)
+        // - REVISIONS : for commit roots (root for branch commits)
+        for (String property : filter(doc.keySet(), COMMITROOT_OR_REVISIONS)) {
+            Map<Revision, String> valueMap = doc.getLocalMap(property);
+            for (Map.Entry<Revision, String> entry : valueMap.entrySet()) {
+                Revision rev = entry.getKey();
+
+                if (!includedClusterIds.isEmpty() && !includedClusterIds.contains(rev.getClusterId())) {
+                    // if the normal sweep(1) happened for some clusterIds already,
+                    // then sweep2 is only done for those.
+                    // In this case includedClusterIds contains that list.
+                    // If that list is empty, it's done for all.
```

Comment:
This is confusing and counter-intuitive to me. In which situation would the list be empty and all clusterIds should we swept?

Reply:
Agree, it is confusing. I have clarified the intention in [this commit](https://github.com/apache/jackrabbit-oak/pull/247/commits/f43e4617fba99ff6e2128cf67e83b27e5dd64c40) : basically the empty list case is only used for testing (since a lot of tests were written before that logic was introduced it was easiest to do it that way)